### PR TITLE
Add data augmentation with modes: none, light, medium, heavy

### PR DESCRIPTION
**Pull request by robertklee**
Adds data augmentation that consists of:

- Blur with probability 30% with sigma
- Horizontal flip (left to right) with probability
- Up to 3 of the following:
    - Sharpening
    - Hue and saturation change of
    - One of:
        - Brightness change of
        - Contrast change of
    - One of:
        - Pixel dropout up to
        - Gaussian noise up to
- Scaling
- Rotation

They've been pretty extensively validated but I haven't run a unit test on the entire validation data generator to see if any keypoints are out of bounds.

## TODO
It would be nice to apply crop and rotate on the original image (before bounding box transformations) so we don't get a black border

### CODE REVIEW

**Review by Ckoelewyn — APPROVED** [review]
It will be interesting to see the different augmentation level's effects on results

File: data_augmentation.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    random_order=True # apply the augmentations in random order
+    )
+
+    ## We may need this line so our pipeline will apply the same operations to keypoints as well as images
+    ## Disabled because we will just pass both the keypoints and the image to the seq pipeline
+    ## which eliminates the possibility that different augmentations will be applied on the 
+    ## original image and keypoints 
+    # aug_pipeline_det = aug_pipeline.to_deterministic()
+
+    return aug_pipeline
```
Comment (robertklee) [inline data_augmentation.py]: I'm working to refactor this code into something that's easier to read. Stay tuned.

File: data_augmentation.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        kpsois_aug.append(kpsoi_aug)
+
+    ia.imshow(
+        np.hstack([kpsois_aug[i].draw_on_image(images_aug[i], size=7) for i in range(len(images_aug))])
+    )
+
+    # TODO need to check if keypoint is out of image is_out_of_image(image_aug)
+
+    # We _could_ use kpsois_aug[i].clip_out_of_image() to remove all keypoints outside of bounds,
+    # but we need keypoint order to manually set heatmaps to 0
```
Comment (robertklee) [inline data_augmentation.py]: rm

File: data_augmentation_tests.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# Test the generator
+import time
+import matplotlib.pyplot as plt
+import pylab
+pylab.rcParams['figure.figsize'] = (10.0, 10.0)
+
+from submodules.HeatMap import HeatMap # https://github.com/LinShanify/HeatMap
+
+start = time.time()
+X_batch, y_stacked, kps_batch = generator[238]
```
Comment (robertklee) [inline data_augmentation_tests.py:59]: TODO: validate that all generator entries have valid keypoints (not out of bounds)

File: hourglass.py [hunk 4]
```diff
@@ -123,6 +134,9 @@ def resume_train(self, batch_size, model_save_base_dir, model_json, model_weight
         print('Loss function selected: {}'.format(loss_str))
         self.loss = get_loss_from_string(loss_str)
 
+        print('Image augmentation strength selected: {}'.format(image_aug_strength))
+        self.img_aug_strength = get_strength_enum_from_string(image_aug_strength)
```
Comment (julianrocha) [inline hourglass.py]: I think it would be more consistent with the other params if `image_aug_strength` and `loss` get passed through to `_start_train` and `_compile_model`in the same way `batch_size` does instead of making instance variables that get initialized at some point after init is called.

File: hourglass.py [hunk 5]
```diff
@@ -123,6 +134,9 @@ def resume_train(self, batch_size, model_save_base_dir, model_json, model_weight
         print('Loss function selected: {}'.format(loss_str))
         self.loss = get_loss_from_string(loss_str)
```
Comment (julianrocha) [inline hourglass.py:135]: I may be missing something but I'm having a hard time finding where this loss value gets used.

Comment (robertklee) [reply hourglass.py:135]: in `compile model`

File: hourglass.py (hunk 4 above)
Comment (robertklee) [reply hourglass.py]: yep good call. It was getting confusing for me too haha. I'll change that.

File: data_generator.py [hunk 6]
```diff
@@ -6,14 +6,18 @@
...
+from imgaug import augmenters as iaa
+from imgaug.augmentables.kps import Keypoint, KeypointsOnImage
 from PIL import Image
 from tensorflow.keras.utils import Sequence
 
 from constants import *
+from data_augmentation import *
 
 # TODO: Data Augementation:
 # - bounding box varied through data augmentation 110% to 150%
```
Comment (julianrocha) [inline data_generator.py:23]: remove todo

File: data_generator.py [hunk 7]
```diff
@@ -111,6 +121,49 @@ def generate_heatmaps(self, label):
...
+            if label[label_idx + (NUM_COCO_KP_ATTRBS-1)] == 0:
+                # invalid keypoint
+                valid[i] = 0
+                kps.append(Keypoint(x=invalid_xy, y=invalid_xy))
+                continue
+            kpx = int(label[label_idx])
+            kpy = int(label[label_idx + 1])
+            kps.append(Keypoint(x=kpx, y=kpy))
+
+        return kps, valid
```
Comment (julianrocha) [inline data_generator.py:140]: lgtm

File: data_generator.py [hunk 8]
```diff
@@ -111,6 +121,49 @@ def generate_heatmaps(self, label):
...
+
+    def convert_imgaug_kpsoi_to_coco_kp(self, kpsoi_aug, valid, image_aug):
+        transformed_label = []
+
+        # Keypoint is visible
+        VALID_KP_V = 1
```
Comment (julianrocha) [inline data_generator.py]: any reason why this is capitalized like the constants?

File: data_generator.py [hunk 9]
```diff
@@ -111,6 +121,49 @@ def generate_heatmaps(self, label):
...
+            if (not valid[i]) or kp.is_out_of_image(image_aug):
+                x, y, v = (0, 0, 0)
+            else:
+                x = kp.x
+                y = kp.y
+                v = VALID_KP_V
+                ## NOTE: Uncomment below if we wish to keep visibility 1,2 flag
+                # label_idx = i * NUM_COCO_KP_ATTRBS  # index for label
+                # visibility = label[label_idx + (NUM_COCO_KP_ATTRBS-1)]
+                # v = visibility
```
Comment (julianrocha) [inline data_generator.py]: in `transform_label` we currently keep v=0 and turn 1 and 2 into 1. So this comment would need to be applied to the previous method

File: data_generator.py [hunk 10]
```diff
@@ -111,6 +121,49 @@ def generate_heatmaps(self, label):
...
+
+            transformed_label.append(x)
+            transformed_label.append(y)
+            transformed_label.append(v)
+
+
+        return np.asarray(transformed_label)
```
Comment (julianrocha) [inline data_generator.py:159]: lgtm

File: data_augmentation.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
+# %% Import required libraries
```
Comment (julianrocha) [inline data_augmentation.py:1]: @robertklee hit me up when the refactor of this file is done

**Review by julianrocha — COMMENTED** [review]
Looks great so far. Let me know when `data_augmentation.py` refactor is done so I can finish the review. Left a few comments. Btw what is `data/Macropus_rufogriseus_rufogriseus_Bruny.jpg` that got added to the repo?

File: data_generator.py (hunk 9 above)
Comment (robertklee) [reply data_generator.py]: Ah, I'll remove then.

**Comment by robertklee** [comment]
@julianrocha that image was for the `data_augmentation_tests.py` input haha

File: data_generator.py (hunk 8 above)
Comment (robertklee) [reply data_generator.py]: no reason haha. removed

File: hourglass.py (hunk 4 above)
Comment (robertklee) [reply hourglass.py]: Fixed

**Review by julianrocha — APPROVED** [review]
nice!
